src/opower/utilities/smud.py

```diff
@@ -82,99 +82,174 @@ def timezone() -> str:
+        #     smud_ssotransition_url,
+        #     headers={"User-Agent": USER_AGENT},
+        #     raise_for_status=True,
+        # )
+        #
+        # SMUD.print_redirects_cookies_response(smud_ssotransition_response, session)
+
+        opower_sso_url = "https://sso.opower.com/sp/ACS.saml2"
+
+        _LOGGER.debug("POSTing opower sso page with SAMLResponse: %s", opower_sso_url)
+
+        opower_sso_response = await session.post(
+            opower_sso_url,
+            data={
+                "SAMLResponse": saml_response,
+                "RelayState": "https://smud.opower.com/ei/app/myEnergyUse",
+            },
+            headers={"User-Agent": USER_AGENT},
+            raise_for_status=True,
+        )
+
+        SMUD.log_response(opower_sso_response, session)
+
+        return
+
+    @classmethod
+    def get_okta_url_from_response_redirect(
+        cls, energyusage_response: ClientResponse
+    ) -> str:
+        """Get the OKTA URL to open next from the last redirect of the previous response."""
+        # https://smud.okta.com/login/sessionCookieRedirect
+        #   ?token=20111...6QJMn
+        #   &redirectUrl=https://smud.okta.com/app/sacramentomunicipalutilitydistrict_opower_1/exk2i...cF0x7/sso/saml
+        #       ?RelayState=https://smud.opower.com/ei/app/myEnergyUse
+        energyUsageResponseRedirectedFinalUrl = energyusage_response.history[-1].url
+
+        query_parts = parse_qs(energyUsageResponseRedirectedFinalUrl.query_string)
+
+        return query_parts["redirectUrl"][0]
+
+    # Store cookies so we can log what is new after each request.
+    cookies = {}
+
+    @staticmethod
+    def log_response(response: ClientResponse, session):
+        """Log any redirects and new cookies. Log full HTML when DEBUG_LOG_RESPONSE is set."""
+        host = response.host  # Is this the request URL or the final redirected url?
+
+        redirects = [r.url for r in response.history]
+        if redirects.__len__() > 0:
+            _LOGGER.debug("Performed %d redirects", redirects.__len__())
+            for redirect in redirects:
+                _LOGGER.debug("-> %s", redirect.__str__())
+
+        if session.cookie_jar.filter_cookies(response.url).__len__() > 0:
+            response_cookie_names = list(
+                session.cookie_jar.filter_cookies(response.url).keys()
+            )
+            last_cookie_names = SMUD.cookies.get(host, [])
+            response_new_cookie_names = set(response_cookie_names) - set(
+                last_cookie_names
+            )
+
+            if response_new_cookie_names.__len__() > 0:
+                _LOGGER.debug(
+                    "Set new cookies: `%s`", "`, `".join(response_new_cookie_names)
+                )
+
+                SMUD.cookies[host] = last_cookie_names + response_cookie_names
+
+        if hasattr("opower", "DEBUG_LOG_RESPONSE") and opower.DEBUG_LOG_RESPONSE:
+            response_html = response.text()
```

Comment:
`response.text()` returns an awaitable, and thus needs to be `response_html = await response.text()`

Reply:
Thanks. That's like the only line I don't have an automated test on and it's wrong!